src/Infrastructure/ServiceBasedPlugin.php

```diff
@@ -0,0 +1,459 @@
+			if ( $service instanceof Deactivateable ) {
+				$service->deactivate( $network_wide );
+			}
+		}
+
+		\flush_rewrite_rules();
+	}
+
+	/**
+	 * Register the plugin with the WordPress system.
+	 *
+	 * @return void
+	 * @throws InvalidService If a service is not valid.
+	 */
+	public function register() {
+		\add_action(
+			static::REGISTRATION_ACTION,
+			[ $this, 'register_services' ]
+		);
+	}
+
+	/**
+	 * Register the individual services of this plugin.
+	 *
+	 * @throws InvalidService If a service is not valid.
+	 *
+	 * @return void
+	 */
+	public function register_services() {
+		// Bail early so we don't instantiate services twice.
+		if ( count( $this->service_container ) > 0 ) {
+			return;
+		}
+
+		// Add the injector as the very first service.
+		$this->service_container->put(
+			static::SERVICE_PREFIX . static::INJECTOR_ID,
+			$this->injector
+		);
+
+		$services = $this->get_service_classes();
+
+		if ( $this->enable_filters ) {
+			/**
+			 * Filter the default services that make up this plugin.
+			 *
+			 * This can be used to add services to the service container for
+			 * this plugin.
+			 *
+			 * @param array<string> $services Associative array of identifier =>
+			 *                                class mappings. The provided
+			 *                                classes need to implement the
+			 *                                Service interface.
+			 */
+			$services = \apply_filters(
+				static::HOOK_PREFIX . static::SERVICES_FILTER,
+				$services
+			);
+		}
+
+		foreach ( $services as $id => $class ) {
+			$id    = $this->maybe_resolve( $id );
+			$class = $this->maybe_resolve( $class );
+
+			// Allow the services to delay their registration.
+			if ( is_a( $class, Delayed::class, true ) ) {
+				$registration_action = $class::get_registration_action();
+
+				if ( did_action( $registration_action ) ) {
+					$this->register_service( $id, $class );
+
+					continue;
+				}
+
+				\add_action(
+					$class::get_registration_action(),
+					function () use ( $id, $class ) {
+						$this->register_service( $id, $class );
+					}
+				);
+
+				continue;
+			}
+
+			$this->register_service( $id, $class );
+		}
+	}
+
+	/**
+	 * Register a single service.
+	 *
+	 * @param string $id ID of the service to register.
+	 * @param string $class Class of the service to register.
+	 */
+	protected function register_service( $id, $class ) {
+		// Only instantiate services that are actually needed.
+		if ( is_a( $class, Conditional::class, true )
+			&& ! $class::is_needed() ) {
+			return;
+		}
+
+		$service = $this->instantiate_service( $class );
+
+		$this->service_container->put( $id, $service );
+
+		if ( $service instanceof Registerable ) {
+			$service->register();
+		}
+	}
+
+	/**
+	 * Get the service container that contains the services that make up the
+	 * plugin.
+	 *
+	 * @return ServiceContainer Service container of the plugin.
+	 */
+	public function get_container() {
+		return $this->service_container;
+	}
```

Comment:
This is what provides external access to the container.

Right now, there's no convenience functionality for this. So to retrieve a service from the container, this would be the complete syntax:
```php
use AmpProject\AmpWp\AmpWpPluginFactory;

$service = AmpWpPluginFactory::create()->get_container()->get( $service_name );
```

If this is wanted, we can add a convenience layer on top of this to provide a syntax like this, for example:
```php
use AmpProject\AmpWp;

$service = Services::get( $service_name );
```

I've not added this just yet, as we should strive to keep the amount of times we directly access the service container to a minimum. Doing so is considered an anti-pattern, namely the "Service Locator" pattern.